packages/job-manager/src/batch/pool-load-generator.ts

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { injectable } from 'inversify';
+
+export interface PoolLoad {
+    activeTasks: number;
+    runningTasks: number;
+}
+
+export interface PoolMetricsInfo {
+    id: string;
+    maxTasksPerPool: number;
+    load: PoolLoad;
+}
+
+export interface PoolMetricsState {
+    poolLoad: PoolLoad;
+    tasksIncrementCount: number;
+    processingSpeed: number;
+    activeToRunningTasksRatio: number;
+    samplingIntervalInSeconds: number;
+    timestamp: number;
+}
+
+@injectable()
+export class PoolLoadGenerator {
+    private lastPoolState: PoolMetricsState;
+
+    public get processingSpeedPerMinute(): number {
+        return Math.round((60 / this.lastPoolState.samplingIntervalInSeconds) * this.lastPoolState.processingSpeed);
+    }
+
+    public get samplingIntervalInSeconds(): number {
+        return this.lastPoolState.samplingIntervalInSeconds;
+    }
+
+    public get activeToRunningTasksRatio(): number {
+        return this.lastPoolState.activeToRunningTasksRatio;
+    }
+
+    public setLastTasksIncrementCount(lastTasksIncrementCount: number): void {
+        this.lastPoolState.tasksIncrementCount = lastTasksIncrementCount;
+    }
+
+    public getTasksIncrementCount(poolMetricsInfo: PoolMetricsInfo, targetActiveToRunningTasksRatio: number): number {
+        // No last pool state available. Use maximum pool capacity multiplied by active to running tasks ratio tasks increment value.
+        if (this.lastPoolState === undefined) {
+            // account for existing active tasks in a pool
+            const initialTasksIncrementCount =
+                poolMetricsInfo.maxTasksPerPool * targetActiveToRunningTasksRatio - poolMetricsInfo.load.activeTasks;
+            this.lastPoolState = {
+                poolLoad: poolMetricsInfo.load,
+                tasksIncrementCount: initialTasksIncrementCount,
+                processingSpeed: 0,
+                activeToRunningTasksRatio: targetActiveToRunningTasksRatio,
+                samplingIntervalInSeconds: 1,
+                timestamp: process.hrtime()[0],
+            };
+
+            return initialTasksIncrementCount;
+        }
+
+        // Calculate processing speed since last state
+        const processingSpeed =
+            this.lastPoolState.tasksIncrementCount + this.lastPoolState.poolLoad.activeTasks - poolMetricsInfo.load.activeTasks;
+
+        // Adapt the initially provided ratio value in case when pool processing speed
+        // is faster than requested active to running tasks ratio.
+        const activeToRunningTasksRatio =
+            poolMetricsInfo.load.activeTasks === 0 && this.lastPoolState.tasksIncrementCount > 0
+                ? this.lastPoolState.activeToRunningTasksRatio + 1
```

Comment:
>activeToRunningTasksRatio [](start = 37, length = 25)

we never seem to be decreasing the ratio.

Reply:
updated